src/cluster.c

```diff
@@ -3525,14 +3525,165 @@ void clusterHandleManualFailover(void) {
  * CLUSTER cron job
  * -------------------------------------------------------------------------- */
 
+/* Check if the node is disconnected and re-establish the connection.
+ * Also update a few stats while we are here, that can be used to make
+ * better decisions in other part of the code. */
+int clusterNodeCronHandleReconnect(clusterNode *node, mstime_t handshake_timeout, mstime_t now) {
+    /* Not interested in reconnecting the link with myself or nodes
+        * for which we have no address. */
+    if (node->flags & (CLUSTER_NODE_MYSELF|CLUSTER_NODE_NOADDR)) return true;
+
+    if (node->flags & CLUSTER_NODE_PFAIL)
+        server.cluster->stats_pfail_nodes++;
+
+    /* A Node in HANDSHAKE state has a limited lifespan equal to the
+        * configured node timeout. */
+    if (nodeInHandshake(node) && now - node->ctime > handshake_timeout) {
+        clusterDelNode(node);
+        return 1;
+    }
+
+    if (node->link == NULL) {
+        clusterLink *link = createClusterLink(node);
+        link->conn = server.tls_cluster ? connCreateTLS() : connCreateSocket();
+        connSetPrivateData(link->conn, link);
+        if (connConnect(link->conn, node->ip, node->cport, server.bind_source_addr,
+                    clusterLinkConnectHandler) == -1) {
+            /* We got a synchronous error from connect before
+                * clusterSendPing() had a chance to be called.
+                * If node->ping_sent is zero, failure detection can't work,
+                * so we claim we actually sent a ping now (that will
+                * be really sent as soon as the link is obtained). */
+            if (node->ping_sent == 0) node->ping_sent = mstime();
+            serverLog(LL_DEBUG, "Unable to connect to "
+                "Cluster Node [%s]:%d -> %s", node->ip,
+                node->cport, server.neterr);
+
+            freeClusterLink(link);
+            return 0;
+        }
+        node->link = link;
+    }
+    return 0;
+}
+
+/* Resize the send buffer of a node if it is wasting
+ * enough space. */
+int clusterNodeCronResizeQueryBuffer(clusterNode *node) {
+    /* If available space is more than twice the size of the buffer then free up unused space */
+    if (node->link != NULL && sdsavail(node->link->sndbuf) / 4 > sdslen(node->link->sndbuf)) {
+        node->link->sndbuf = sdsRemoveFreeSpace(node->link->sndbuf);
+    }
+    return 0;
+}
+
+/* Check if we need to flag the node as failing.
+ * This function is also responsible to:
+ * 1) Check if there are orphaned masters (masters without non failing
+ *    slaves).
+ * 2) Count the max number of non failing slaves for a single master.
+ * 3) Count the number of slaves for our master, if we are a slave. */
+int clusterNodeCronCheckFailingNodes(clusterNode *node, int* update_state, int* orphaned_masters, int* max_slaves, int* this_slaves) {
```

Comment:
Ok, I take back what I said earlier. This function doesn't really feel coherent, and we should probably spend more effort to refactor this. My new proposal would be to refactor the code to have one loop to do the resizeBuffer + handle reconnect, which seem consistent on their own, and then leave the failing node check alone.

Reply:
yeah, this function is different because it has to update all these counters. I can move this back to its own loop if that is what people want.

Reply:
Yeah, I think coherence is important here unless we've identified it as a performance gap.

Reply:
idk what coherence means, but i'll change it. If we keep the resizeBuffer + handle reconnect then it has 2 for loops through the nodes the same as before my changes. I'll leave the rest of the functions as is and move `clusterNodeCronCheckFailingNodes` back to the main cron function.

Reply:
Sounds good